Add Caution about undelegation flaw (#3856)
w3f/polkadot-wiki: docs/maintain/maintain-guides-democracy.md

## Undelegate a Vote

:::caution

> Currently, the undelegation logic has a flaw in it. If you choose to delegate and undelegate two

Reviewer: In the scenario you mentioned, where someone delegates and undelegates, and then delegates and undelegates, it would not be a flaw to reset the lock period to the second undelegation.

I think the issue outlined is that the user first reduces the number of DOT they delegate. Say a user is delegating 500 DOT and wanted to reduce it to 200 DOT. They can issue an extrinsic to do that and the 300 DOT should start unlocking. During the period of unlock, the user chooses to undelegate the current voting power of 200 DOT. This resets the unlock period of the 300 DOT that was unlocking in the background.

@michalisFr Is my understanding correct?

Reviewer 2: Technically, when you reduce your delegation, the whole of the initial amount is set to unlock. If then, after the unlock period, you remove the lock, only the difference will become transferable because there is the lock of the current delegation.

So, if you delegate 500 DOT and then change it to 200 DOT, the lock set to expire is for the 500 DOT. When that lock expires and you remove it, 300 DOT will become transferable, because there is still a lock in existence for 200 DOT. It's not that when you do this you unlock only 300 DOT. 

The issue is that if you do this (the simplest case):
1. Delegate 500 DOT
2. Delegate 200 DOT at time t -> This creates a lock for 200 DOT and sets the lock for 500 DOT to expire at t + unlocking period`. So, the "overall" lock is for 500 DOT.
3. Undelegate at `t1 < t + unlocking period` (i.e. before the previous lock expired) -> Resets the lock for 500 DOT to expire at `t1 + unlocking period`

The expected behaviour would be that 500 DOT still unlocks at t + unbonding period and the 200 DOT at t1 + unbonding period.

Obviously, this only affects accounts that **delegated with a conviction**. 

Also, I wouldn't call this a "flaw". It's an edge case that wasn't considered.

Author: > The formatting for the caution is off. Check https://docusaurus.io/docs/markdown-features/admonitions for the correct way to use admonitions.
> 
> <img alt="image" width="910" src="https://user-images.githubusercontent.com/86818441/193341628-b3c49826-f5ca-40d6-8adf-b87ba36f0d92.png">

Ah, this seems to be an issue with prettier:
If the closing colons are on the next line, it puts it at the end of the sentence.

## Undelegate a Vote

:::caution

If there is an existing lock due to a previous delegation change or undelegation, any new change or
undelegation will restart the lock period for the larger DOT amount and the longest conviction
period between the existing and the new lock.

> This will only matter to accounts with conviction, as no conviction doesn't instill a lock period.

Reviewer: ```suggestion
This will only matter to accounts with conviction, as the accounts with no conviction don't have any lock period.
```

Examples:

1. Delegate 500 DOT with 1x conviction, then change delegation to 1000 DOT with 1x conviction, the
   lock period will reset for 1000 DOT with 1x conviction.

2. Delegate 500 DOT with 3x conviction, then change the delegation to 1000 DOT with 1x conviction,
   the lock period will reset for 1000 DOT with 3x conviction.

3. Delegate 500 DOT with 1x conviction, then change the delegation to 200 DOT with 1x conviction,
   the lock period will reset for 500 DOT with 1x conviction.

To understand this in further detail checkout

> [this](https://substrate.stackexchange.com/questions/5067/delegating-and-undelegating-during-the-lock-period-extends-it-for-the-initial-am)
> stackexchange post.

Reviewer: ```suggestion
[this stackexchange post.](https://substrate.stackexchange.com/questions/5067/delegating-and-undelegating-during-the-lock-period-extends-it-for-the-initial-am)
```
